common/variant/src/main/java/org/apache/spark/types/variant/VariantUtil.java

```diff
@@ -392,21 +392,32 @@ public static double getDouble(byte[] value, int pos) {
     return Double.longBitsToDouble(readLong(value, pos + 1, 8));
   }
 
+  // Check whether the precision and scale of the decimal are within the limit.
+  private static void checkDecimal(BigDecimal d, int maxPrecision) {
+    if (d.precision() > maxPrecision || d.scale() > maxPrecision) {
```

Comment:
shall we also check `precision > scale`?

Reply:
`precision >= scale` is a requirement in Spark, not in variant spec. In the new test case I add, `0.01` has a precison of 1 and a scale of 2. It is a valid decimal in the variant spec but requires some special handling in `schema_of_variant`.